internal/dcache/rpc/client/client_pool.go

```diff
@@ -193,6 +235,88 @@ func (cp *clientPool) resetRPCClient(client *rpcClient) error {
 	return nil
 }
 
+// Close client and create a new one to the same target/node as 'client'.
+// This is typically used when a node goes down and comes back up, rendering all existing clients
+// "broken", they need to be replaced with a brand new connection/client.
+func (cp *clientPool) resetRPCClient(client *rpcClient) error {
+	return cp.resetRPCClientInternal(client, true /* needLock */)
+}
+
+// Reset all connections in the client pool corresponding to 'client'. This client would have been allocated
+// using a prior call to getRPCClient().
+// This is used for draining and fixing all old/bad connections in the connection pool in case of the target
+// node restarting. It closes the passed in connection and all existing connections in the pool and replaces
+// them with newly created ones.
+func (cp *clientPool) resetAllRPCClients(client *rpcClient) error {
+	//
+	// resetAllRPCClients() will be called only when we know for sure that an RPC request made using 'client'
+	// failed with a "connection reset by peer" error, we reset that client and all others in the pool.
+	//
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	numConnReset := 0
+	ncPool, exists := cp.clients[client.nodeID]
+
+	// client is allocated from the pool, so pool must exist.
+	common.Assert(exists)
+	common.Assert(ncPool.nodeID == client.nodeID, ncPool.nodeID, client.nodeID)
+
+	//
+	// Reset this client. This closes this client, creates a new one and adds it to the pool.
+	// It can only fail if thrift fails to create a new connection. This can happen when a node
+	// (rather the blobfuse service in the node) has gone down.
+	//
+	err := cp.resetRPCClientInternal(client, false /* needLock */)
+	if err != nil {
+		err = fmt.Errorf("Failed to reset RPC client to %s node %s: %v",
+			client.nodeAddress, client.nodeID, err)
+		log.Err("clientPool::resetAllRPCClients: %v", err)
+		// Connection refused is the only viable error. Assert to know if anything else happens.
+		common.Assert(rpc.IsConnectionRefused(err))
+		return err
+	}
+
+	numConnReset++
+
+	//
+	// Clients present in the pool If it's less than cp.maxPerNode, rest are currently allocated to other
+	// callers. We cannot replenish those. Those will be reset by their respective caller when their RPC
+	// requests fail with "connection reset by peer" error.
+	//
+	numClients := len(ncPool.clientChan)
+	common.Assert(numClients <= int(cp.maxPerNode), numClients, cp.maxPerNode)
+
+	//
+	// Reset all remaining clients in the pool. We try to reset as many as we can, and don't fail
+	// on error, as we have reset at least one client.
+	//
+	for i := 0; i < numClients; i++ {
+		client, err = cp.getRPCClientNoLock(client.nodeID)
+		// getRPCClientNoLock should not fail, because we have the clientPool for this client.
+		common.Assert(err == nil, err)
+
+		err = cp.resetRPCClientInternal(client, false /* needLock */)
```

Comment:
This would again reset the client which we had already done in L270.
```
        //
	// Reset this client. This closes this client, creates a new one and adds it to the pool.
	// It can only fail if thrift fails to create a new connection. This can happen when a node
	// (rather the blobfuse service in the node) has gone down.
	//
	err := cp.resetRPCClientInternal(client, false /* needLock */)
```